dali/operators/input/video_input.cc

```diff
@@ -0,0 +1,149 @@
+#include <memory>
+#include <vector>
+
+namespace dali {
+
+/// By definition, the input batch size of this Operator is always 1.
+constexpr int input_batch_size = 1;
+
+
+template<>
+bool VideoInput<CPUBackend>::SetupImpl(std::vector<OutputDesc> &output_desc,
+                                       const Workspace &ws) {
+  if (!valid_) {
+    InputOperator<CPUBackend>::HandleDataAvailability();
+    TensorList<CPUBackend> encoded_videos;
+    frames_decoders_.resize(input_batch_size);
+    auto &thread_pool = ws.GetThreadPool();
+    this->ForwardCurrentData(encoded_videos, thread_pool);
+
+    // Creating FramesDecoders
+    auto sample = encoded_videos[0];
+    auto data = reinterpret_cast<const char *>(sample.data<uint8_t>());
+    size_t size = sample.shape().num_elements();
+    frames_decoders_[0] = std::make_unique<FramesDecoder>(data, size, false);
+
+    assert(output_descs_.empty());
+    DetermineOutputDescs(static_cast<int>(frames_decoders_[0]->NumFrames()));
+
+    // This has to be done for every video file, since we need to know the shape of the frames.
+    if (last_sequence_policy_ == "pad") {
+      InitializePadValue(0);
+    }
+
+    valid_ = true;
+  }
+  output_desc.resize(1);
+  output_desc[0] = output_descs_.front();
+  output_descs_.pop_front();
+  return true;
+}
+
+
+template<>
+void VideoInput<CPUBackend>::RunImpl(Workspace &ws) {
+  auto &output = ws.Output<CPUBackend>(0);
+  bool full_sequence;
+  for (int64_t s = 0; s < output.num_samples(); s++) {
+    auto pad_value =
+            last_sequence_policy_ == "pad" ? std::optional<SampleView<CPUBackend>>(GetPadFrame())
+                                           : std::nullopt;
+    full_sequence = DecodeFrames(output[s], 0, frames_per_sequence_, pad_value);
+    if (!full_sequence) {
+      break;
+    }
+  }
+  if (!full_sequence || frames_decoders_[0]->NextFrameIdx() == -1) {
+    Invalidate();
+  }
+}
+
+
+DALI_SCHEMA(experimental__inputs__Video)
+                .DocStr(
+                        R"code(
+Decodes a video from a memory buffer. Returns a batch of sequences of frames, where
+each frame has ``(F, H, W, C)`` layout (where ``F`` is the number of frames in a sequence).
+
+When using ``fn.inputs.video`` operator inside the DALI Pipeline, user needs to provide the data
+using :meth:`Pipeline.feed_input`. When the Operator is fed with data, the Pipeline can be ran
+multiple times and the ``fn.inputs.video`` operator will return consecutive sequences, as long as
+there is enough data remaining. When the source of the frames (the video file) depletes, user needs
+to call another ``feed_input`` to provide the next video file to the operator. This Operator has an
+inner-queue for the data, so the ``feed_input`` may be called multiple times and when given video
+file ends, the Operator will fetch the next one automatically from the top of the queue.
+Running the pipeline while there is no data for the ``fn.inputs.video`` to run results in an error.
```

Comment:
Is there a mechanism (exception based or otherwise) to tell the user we ran out of data without erroring?

Reply:
There's not. I don't know if we need one. I might be wrong, but I believe that when you're using ExternalSource with `feed_input`, so do not have such mechanism either, is that correct? If so and if we need it, we might design something generic.

For sure, adding such a thing would be rather a bigger effort. ExternalSource has the `cycle` parameter, but the whole operator is partially implemented in Python, therefore it's easier to throw a `StopIteration` there. With `VideoInput` I'm afraid it won't be that simple.

Reply:
So, does the user need to know how many times we can call run until the next feed_input needs to be called? I was just wondering how the usage will look in practice

Reply:
In the C API we will have the `EventId` mechanism (name TBD). This is the way user will know when to call `feed_input`. In the Python API, it is not decided yet. I mean, it may and probably will be the `EventId` mechanism also, but this is a different and big effort to actually add it to the Python API. For now we're assuming that user knows how many frames the video has and therefore knows when to call `feed_input`. That's, BTW, how I wrote the tests.